[Bibi/refactor] DIContainer 및 Coordinator 적용 및 이슈목록 갱신 문제 디버깅

bibi6666667 opened the pull request:
안녕하세요, Bibi입니다!
프로젝트 리팩토링을 하며 목표로 했던 아래 사항들을 완료해 PR 드립니다.
- DIContainer 적용하기
- Coordinator 패턴 적용하기
- 새 이슈 생성 이후, 생성된 이슈를 반영해 이슈목록 화면에 보여주기

---

## 이슈목록 갱신에 대한 질문

### 제가 의도한 동작

1. 이슈 생성 화면에서 이슈 생성 버튼을 누른다.
2. 정상적으로 생성되면 이전 화면인 이슈 목록 화면으로 돌아간 다음,
3. 이슈 목록 화면을 갱신해 새로 생성된 이슈를 화면에 즉시 보여준다.

### 제가 생각한 흐름

1. NewIssueVC의 createButton 터치 시, 입력된 정보를 바탕으로 model을 통해 Issue Create API로 요청 전송 (비동기)
2. 이슈가 잘 생성된 경우 coordinator를 통해 이전 화면으로 돌아감 (pop)
3. 이전 화면인 IssueVC 에서 coordinator를 통해 이슈목록을 다시 조회 하도록 API 요청 전송 (비동기)
4. 받아온 이슈목록을 가지고 화면을 갱신

### 제가 한 처리

(fetch = API에 이슈목록 요청,  reload = collectionView.reloadData() 입니다!)
- IssueVC를 생성할 때 IssueModel을 만들어 넣어줌
- viewDidLoad()에서 model의 issues에 변경이 있을 때 실행될 메서드를 정의해 넣어줌
- 처음 이슈목록을 보여줄 때는 fetch 후 응답이 오면 reload 진행
- 새 이슈 생성이 성공하면, 새로 생성된 이슈를 화면에 반영하기 위해 IssueVC로 돌아와 fetch 후 응답이 오면 reload 진행 
  - 데이터 fetch를 비동기 요청으로 보냈는데, 응답이 오지 않은 상태에서 reload했음

### 문제 분석

- 새 이슈 생성 후 돌아왔을 때 다시 요청하면, 새로 생성된 이슈가 오지 않음
- 포스트맨에서 확인 결과, 시뮬레이터에서 새 이슈 생성 요청, 완료 이후 즉시 이슈목록 조회를 했더니 새 이슈가 반영되지 않을 때도 있었음

- 새 이슈 생성 요청 -> 완료 이후, 이슈목록 조회했을 때 새 이슈가 생성되는 시간이 조금 걸리는 것으로 추정
  - 1. 새 이슈 생성 이후부터 이슈목록 조회 사이에, 적당한 시간간격을 두고 조회해 보기로 함
  - 2. 새로 생성하는 이슈의 제목을 가지고 있다가, 조회한 이슈목록에 해당 이슈가 포함되어 있지 않다면 재요청하는 방법도 있을 것 같다.

### 해결(?)

- 1, 2 의 방법을 모두 사용해 봤는데, 이렇게 하니 적당한 시간동안 서버에 올라갈 때 까지 기다려 주면서 새 이슈가 화면에 반영될 때까지 재요청을 할 수 있었다.
  - 다만 이렇게 요청하면 API서버에 부하가 걸리진 않을지..가 조금 걱정입니다.🤔

## DIContainer 관련

- container에 register하는 시점은 언제가 좋은지에 대한 고민 - 각 객체마다 생성 시점이 달라서 등록 시점도 다를 수 밖에 없는 것일까요
- resolve할 때 마다 옵셔널바인딩을 하는데, 없을 때 처리를 어떻게 하는 게 좋을지 고민했습니다
    - 그냥 `guard let issueModel: IssueModel = container.resolve() else { return }` 하고 있는데, 모델이 없을 때 아무 동작도 하지 않게 된다는 문제
- 동적으로 생성되는 VC는 container를 사용할 수 없는 건지
    - 예를 들어 : IssueVC는 한번 만들어 두고 계속 쓰는 게 아니라 selectedRepo가 달라질때마다 새로 만들어야 하는데.. 그럼 컨테이너로 관리하기 적당하지 않은 대상인가? 라는 의문 발생
    - 우선은 Container를 쓰지 않고 Coordinator에서 필요할 때 직접 생성해 사용하는 방식으로 구현했습니다
- register할 때 종속성이 발생하는 문제
    - 예를 들어 : IssueVC를 생성하려면 IssueModel이 필요한데, 모델과 뷰컨을 각각 만들고 넣어 주어야 하는지, 아니면 뷰컨을 만들 때 모델을 만들어 넣어 주어야 하는지 고민했습니다

## Coordinator 관련

- 뷰컨이 사라질 때 childCoordinator 목록 관리
    - 현재는 fromViewCoordinator(`UINavigationController`)를 각 뷰컨으로 캐스팅이 되는지 판단 후 지워주고 있습니다 - 이로 인해 비슷한 코드가 대량으로 발생하는 문제

bibi6666667 committed:
Feat : AppDelegate를 위한 코디네이터인 AppCoordinator 생성 및 실행 시점 rootVC 판단 로직 분리

bibi6666667 committed:
Feat: Container에 register, resolve 패턴 적용해 Login/Repos 화면 전환

bibi6666667 committed:
Feat: ReposVC에 코디네이터와 컨테이너 적용

bibi6666667 committed:
Feat: ReposViewController를 보여줄 때 ReposCoordinator를 사용하도록 변경

bibi6666667 committed:
Feat: IssueViewController 화면 흐름을 관리하는 IssueCoordinator 추가 #28

bibi6666667 committed:
Feat: NewIssueViewController를 관리하는 NewIssueCoordinator 추가 및 새 이슈 생성에 필요한 NewIssueFormat 구조체 추가

bibi6666667 committed:
Feat: 모든 뷰컨에 coordinator 프로퍼티 추가 및 NewIssue 생성 시 뒤로가기 동작 구현 #28

bibi6666667 committed:
Feat: OptionSelectViewController를 관리하는 OptionSelectCoordinator 추가 #28

bibi6666667 committed:
Feat : coordinator생성 시 register한 뒤, childCoordinators 관리할 때 활용 #28

기존 VC에서 coordinator를 직접 소유하던 부분을 삭제하고
AppCoordinator에서 childCoordinators를 관리할 때 register된 coordinator 사용

bibi6666667 committed:
Feat : 이슈생성, 옵션선택 후 coordinator에서 이전 화면으로 되돌아가며 reload호출 (동작하지 않음) #28

bibi6666667 committed:
Feat : OptionSelectVC를 coordinator로 연결 - 옵션마다 다른 화면 생성 #28

bibi6666667 committed:
Merge branch 'bibi/refactor' of https://github.com/Jinsujin/issue-tracker into bibi/refactor

# Conflicts:
#	IssueTracker/IssueTracker/AppDelegate.swift
#	IssueTracker/IssueTracker/Issue/IssueViewController.swift
#	IssueTracker/IssueTracker/NewIssue/NewIssueViewController.swift
#	IssueTracker/IssueTracker/Repos/ReposViewController.swift

bibi6666667 committed:
Feat : 이슈 생성 요청과 이슈목록 요청 사이에 딜레이 추가, 받아온 이슈에 새로 생성한 이슈가 없다면 재요청 #28

bibi6666667 committed:
Chore : 미사용 메서드와 주석 삭제 #28

bibi6666667 committed:
Chore : 불필요한 주석 및 코드 삭제 #28

GangWoon commented on IssueTracker/IssueTracker/Coordinator/AppCoordinator.swift:
```diff
@@ -0,0 +1,189 @@
+        
+        // navStack에 존재하지 않음 = pop됨 : 해당 뷰컨의 coordinator를 childCoordinators에서 지워야 함
+        // MARK: 모든 뷰컨의 코디네이터 관리하면서도 코드중복 피하는 방법..?
+        if fromViewController as? LoginViewController != nil,
```
is라는 비교문도 존재합니다만, 해당 부분에서 수동적인 작업이 많이 발생하는 듯한 느낌을 받게 됩니다.
화면이 증가할때마다 해당 구분이 점점 비대해지지 않으까요?

GangWoon commented on IssueTracker/IssueTracker/Coordinator/AppCoordinator.swift:
```diff
@@ -0,0 +1,189 @@
+import Foundation
+import UIKit
+
+class AppCoordinator: NSObject, Coordinator {
```
코디네이터와 컨테이너의 역활에 대해서 명확하게 말씀해주실 수 있으신가요?

bibi6666667 replied:
제가 이해한 바는 다음과 같습니다.
컨테이너(DIContainer)
- 앱에서 사용되는 주요 인스턴스를 관리하는 역할
- 인스턴스를 생성할 때 컨테이너에 등록(register)한 뒤, 필요할 때 컨테이너로부터 꺼내(resolve) 사용할 수 있음
- 동일 인스턴스가 여러 곳에서 중복 생성되는 것을 방지

코디네이터
- 모든 뷰 컨트롤러들을 관리해 화면 흐름을 조정하는 역할
- 각 뷰 컨트롤러들을 독립적으로 만들어 줌 - 다른 뷰컨트롤러에 대해 알지 못하게 함
- 화면 흐름을 코디네이터라는 한 곳에서 모아서 관리할 수 있음

GangWoon commented on IssueTracker/IssueTracker/Container.swift:
```diff
@@ -1,100 +1,92 @@
-    // VC를 찍어내는 역할
     
     let environment: ContainerEnvironment
+    private var objects: [String: Any] = [:]
```
무엇을 저장하는 콜랙션일까요?

bibi6666667 replied:
컨테이너에 등록되는 객체들을 저장하는 역할을 하는 컬렉션입니다!
키는 객체의 이름을, 값은 등록된 객체를 넣어 딕셔너리로 저장했습니다.

bibi6666667 replied:
변수명을 registeredObjects로 바꾸어 모호함을 줄여 보겠습니다!

GangWoon commented on IssueTracker/IssueTracker/Coordinator/IssueCoordinator.swift:
```diff
@@ -0,0 +1,75 @@
+    
+    func start() {
+        guard let repo = selectedRepo else {
+            return
```
말씀하신 부분이 어색하다고 느껴지신다면, 예외처리 화면을 추가하셔도 좋을 거 같습니다.

bibi6666667 replied:
음.. 그런 방법이 있군요! 한번 도전해 보겠습니다. 감사합니다🙏

GangWoon commented on IssueTracker/IssueTracker/Coordinator/IssueCoordinator.swift:
```diff
@@ -0,0 +1,75 @@
+        guard let repo = selectedRepo else {
+            return
+        }
+        // MARK: 동적으로 생성되어야 하는 VC는 Container를 사용할 수 없나?
```
해당 코맨트가 무엇을 의미하는지 명확하게 알 수 없었습니다.
반대로 뷰컨트롤러가 영구적으로 존재하게 만든다면 릭이 발생하는 안 좋은 상황이 될 거 같습니다.

GangWoon commented on IssueTracker/IssueTracker/Coordinator/IssueCoordinator.swift:
```diff
@@ -0,0 +1,75 @@
+        guard let viewController = self.viewController else {
+            return
+        }
+        DispatchQueue.main.asyncAfter(deadline: .now() + 4.0) {
```
말씀하신 방법대로 깃헙 api가 작동한다면, 해당 부분에 대해선 제가 생각할 수 있는건 아래와 같습니다.
생성 버튼을 누를때, 인디케이터를 띄우면서 서버에서 정상적으로 성공 응답과, 해당 아이템이 이슈 목록으로부터 추가가 됐을때 ,이전 화면으로 넘어와서 보여주는 방식이 편해보입니다.

bibi6666667 replied:
오.. 그런 방법이 있겠군요! 그게 사용자 입장에서 훨씬 더 자연스러울 것 같습니다. 흐름을 수정해 볼게요 감사합니다🙏

GangWoon commented on IssueTracker/IssueTracker/Coordinator/OptionSelectCoordinator.swift:
```diff
@@ -0,0 +1,60 @@
+    func start() {
+        let optionSelectModelEnvironment = OptionSelectModelEnvironment {
+            [weak self] repo, completion in
+            self?.container.environment.issueService.requestRepositoryLabels(repo: repo, completion: completion)
```
위와 같은 그림을 피하기 위해서 struct 객체를 사용해서 레이어를 두시면 각각 집어넣는 문제를 피하실 수 있습니다.

bibi6666667 replied:
이 부분도 레이어를 쓰면 개선할 수 있군요. 수정해 보겠습니다! 감사합니다👍

GangWoon commented:
```container에 register하는 시점은 언제가 좋은지에 대한 고민 - 각 객체마다 생성 시점이 달라서 등록 시점도 다를 수 밖에 없는 것일까요```

모든 객체를 미리 생성해서, 생성 시점을 맞출 수 있다면, 초반 로딩시간이 너무 길어지지 않을까요? 각 객체마다 필요한 시저에 생성하는게 자연스럽게 느껴집니다.

```
resolve할 때 마다 옵셔널바인딩을 하는데, 없을 때 처리를 어떻게 하는 게 좋을지 고민했습니다
그냥 guard let issueModel: IssueModel = container.resolve() else { return } 하고 있는데, 모델이 없을 때 아무 동작도 하지 않게 된다는 문제
```
이유에 대해서 표현할 수 있다면, 사용자에게 적절한 사용자 경험을 제공할 수 있을 거 같습니다.

```
동적으로 생성되는 VC는 container를 사용할 수 없는 건지
예를 들어 : IssueVC는 한번 만들어 두고 계속 쓰는 게 아니라 selectedRepo가 달라질때마다 새로 만들어야 하는데.. 그럼 컨테이너로 관리하기 적당하지 않은 대상인가? 라는 의문 발생
```
뷰컨트롤러는 특정 모델에 대해서만 의존하게 된다면, 달라지는 모델만 변경시켜주면서 뷰는 계속 재사용할 수 있지 않을까요?
뷰컨트롤러가 해제되지 않는다는 가정하에 의견입니다.


```
우선은 Container를 쓰지 않고 Coordinator에서 필요할 때 직접 생성해 사용하는 방식으로 구현했습니다
register할 때 종속성이 발생하는 문제
예를 들어 : IssueVC를 생성하려면 IssueModel이 필요한데, 모델과 뷰컨을 각각 만들고 넣어 주어야 하는지, 아니면 뷰컨을 만들 때 모델을 만들어 넣어 주어야 하는지 고민했습니다
```
해당 질문에 차이점에 대해서 설명해주실 수 있으신가요?


```
Coordinator 관련
뷰컨이 사라질 때 childCoordinator 목록 관리
현재는 fromViewCoordinator(UINavigationController)를 각 뷰컨으로 캐스팅이 되는지 판단 후 지워주고 있습니다 - 이로 인해 비슷한 코드가 대량으로 발생하는 문제
```
코맨트 드렸듯이 문제가 발생하고 있는 부분같습니다. 어떻게 처리할 수 있을까요? 해당 부분은 고민을 추가적으로 하셨으면 좋겠습니다.

bibi6666667 commented:
`모든 객체를 미리 생성해서, 생성 시점을 맞출 수 있다면, 초반 로딩시간이 너무 길어지지 않을까요? 각 객체마다 필요한 시저에 생성하는게 자연스럽게 느껴집니다.`
- 그러고 보니 초반 로딩시간을 생각하지 못했군요..! 필요할 때 생성해 등록하도록 하겠습니다.

`이유에 대해서 표현할 수 있다면, 사용자에게 적절한 사용자 경험을 제공할 수 있을 거 같습니다.`
- 아하.. 초기화가 왜 안됐는지, 모델이 왜 없었는지를 사용자에게 풀어서 설명해 주면 되나요? 한번 시도해 보겠습니다!

`뷰컨트롤러는 특정 모델에 대해서만 의존하게 된다면, 달라지는 모델만 변경시켜주면서 뷰는 계속 재사용할 수 있지 않을까요?뷰컨트롤러가 해제되지 않는다-는 가정하에 의견입니다.`
이 부분은 같이 공부하시는 분이 제안주셨던 부분이기도 했는데요,
- "모델이 달라지면 결국 뷰도 다시 그려야 하는데, 그럼 뷰를 재사용할 수 없는 것 아닌가?"라는 생각이 들어서 시도하려다가 말았습니다. 예를 들어 repo가 달라지면 IssueModel이 변경되어 IssueViewController도 다시 그려야 하는데, 모델이 변경되기 전에 그려둔 뷰를 '재사용'할 수 있는 건가요..? 결국 바뀐 repo를 가지고 IssueViewController 초기화를 다시 해야 하는 거 아닌가 라는 의문이 들었습니다.
- 현재 뷰컨트롤러에 deinit을 확인해 보니 호출되지 않고 있습니다.. 어디에선가 순환참조가 발생하고 있는 건지, 아니면 Container에 뷰컨트롤러를 register해서 그런 건지 확인해 보겠습니다.

`해당 질문에 차이점에 대해서 설명해주실 수 있으신가요?`
- ReposViewController를 만들 때 ReposModel을 필요로 함
- i) 컨테이너의 registerReposViewController() 내부에서, ReposModel을 먼저 등록하고, 그 모델을 가지고 ReposViewController를 만들어 등록
  - registerReposViewController에서 ReposModel까지 등록하게 되어, 단일책임원칙에 어긋난다는 단점
- ii) 컨테이너의 registerReposModel()에서 registerReposViewController()에서 각각의 객체를 등록함. 이후 registerReposModel() -> registerReposViewController() 순서대로 호출해 ViewController를 생성 (현재 상태)
  - 두 메서드를 반드시 순서대로 호출해야 하며, 호출 순서가 바뀌면 ReposViewController의 초기화가 보장되지 않는다는 단점

`코맨트 드렸듯이 문제가 발생하고 있는 부분같습니다. 어떻게 처리할 수 있을까요? 해당 부분은 고민을 추가적으로 하셨으면 좋겠습니다.`
- 이 부분은 저도 개선이 필요하다고 느꼈는데, 제가 생각한 처리로는 Coordinator-ViewController를 묶어서 관리하다가 fromViewController가 넘어왔을 때 대응되는 Coordinator를 삭제하도록 하면 될 수도 있겠다.. 고 생각했습니다. 혹시 더 나은 처리가 있다면 조언 부탁드립니다!🙇‍♀️

bibi6666667 committed:
refactor: ViewController deinit이 호출되지 않는 문제 디버깅 #28
- delegate 프로토콜을 weak var로 갖도록 변경
- 클로저 내에 선언된 self 를 weak self로 변경

bibi6666667 committed:
refactor: 변수명을 적절하게 수정 #28

bibi6666667 committed:
refactor: 불필요한 self 키워드 삭제 #28

bibi6666667 committed:
refactor: 이슈 생성 요청 이후 새 이슈가 생성되었는지 확인한 다음 이슈 목록으로 되돌아가도록 변경 (뷰 미완성) #28

bibi6666667 committed:
refactor : 새 이슈 생성 -> 이슈 조회 흐름 수정. #28
- 새 이슈 생성 시, "이슈생성 요청" 이후 "이슈목록 조회 요청" 을 보내 생성된 이슈가 반영될때까지 시간 간격을 두고 재요청을 한다. 이 때 indicator를 나타낸다.
- 반영이 되면 이전 화면인 이슈목록으로 돌아가 이슈목록을 조회하고, 화면을 갱신한다.

bibi6666667 committed:
debug : 로그인 직후 앱으로 돌아왔을 때 저장소 목록이 나오지 않는 현상 디버깅(해결) #28

bibi6666667 committed:
refactor : Container에서 뷰컨트롤러와 관련 코디네이터를 묶어 관리하고 삭제 작업도 처리하도록 변경 #28

bibi6666667 committed:
refactor : 뷰컨트롤러와 관련 코디네이터 이름을 담은 ViewControllerCoordinator 열거형을 사용해 중복 코드 개선 #28

bibi6666667 committed:
chore : 불필요한 주석 및 print문 삭제 #28

bibi6666667 committed:
debug : IssueService 초기화를 위해 토큰 기본값을 넘기는 대신 옵셔널로 타입 수정 #28

bibi6666667 commented:
안녕하세요, bibi입니다!

해당 PR에서 조언 주신 주요 개선사항들의 반영이 완료되어 코멘트를 추가합니다🙂

개선된 사항들은 크게 다음과 같습니다.

- viewcontroller의 deinit이 호출되지 않던 이유 디버깅
- 새 이슈 생성 → 이슈 조회의 흐름 수정
- 로그인화면 → 로그인 완료 → 앱으로 돌아왔을 때 로그인 화면에서 멈춰 있는 현상 디버깅
- 자식 코디네이터 관리 방식 수정

### 새 이슈 생성 → 이슈 조회의 흐름 수정
- 새 이슈 생성 요청을 보낸 후,  이전 화면으로 돌아가기 전에 이슈 조회를 요청해 새 이슈가 생성되었는지 확인. 이 때 버튼 상단에 indicator 표시.
- 새 이슈가 생성되었다면, 이전 화면인 이슈 조회 목록으로 돌아가 생성된 새 이슈를 포함해 화면에 표시
- 새 이슈가 아직 생성되지 않았다면, indicator를 계속 나타내며 지정된 시간 간격으로 반복하여 이슈 조회 재요청을 보냄
Timer와 RunLoop에 대해 공부하고, 자연스러운 화면 처리 flow를 알게 되었습니다!

### 자식 코디네이터 관리 방식 수정
- 부모 코디네이터(`AppCoordinator`)에 자식 코디네이터 관리를 위한 단순반복 코드가 계속 필요한 상태에서 개선이 필요하다고 느꼈습니다.
- 코디네이터와 그 코디네이터가 관리하는 쌍을 딕셔너리로 만들고, 그 관리 주체를 AppCoordinator가 아닌 Container로 변경했습니다.
  - 생성된 객체를 관리하는 역할을 하는 곳이 이미 있는데, 굳이 부모 코디네이터에서 '자식 코디네이터들' 외에 '코디네이터-뷰컨트롤러 쌍'까지 관리할 필요는 없다고 생각해 이렇게 옮겨 보았습니다.
  - 뷰컨트롤러를 컨테이너에 register하는 시점에는 그 뷰컨트롤러를 관리하는 코디네이터가 이미 만들어져 있으므로, private 메서드로 register하는 객체가 UIViewController라면 그 쌍이 되는 코디네이터를 함께 딕셔너리에 등록하도록 했습니다.
- "코디네이터-뷰컨트롤러" 쌍을 관리하는 enum을 만들고 `CaseIterable` 프로토콜을 준수하도록 해, 모든 뷰컨트롤러의 타입을 일일히 타입캐스팅을 통해 확인하는 대신 클래스 명만 순회하며 비교하도록 개선했습니다.

### viewcontroller의 deinit이 호출되지 않던 이유 디버깅
`뷰컨트롤러는 특정 모델에 대해서만 의존하게 된다면, 달라지는 모델만 변경시켜주면서 뷰는 계속 재사용할 수 있지 않을까요?뷰컨트롤러가 해제되지 않는다-는 가정하에 의견입니다.`
해당 질문에 답변을 하면서, ‘그러고 보니 뷰컨의 deinit이 호출이 안 되네?’ 라고 느꼈고, 순환참조가 발생하고 있다고 생각해 디버깅을 시작했습니다.
- 참고 문서 : https://www.swiftdevcenter.com/reasons-for-deinit-is-not-getting-called-ios-swift/
- 저의 경우 원인은 1. 뷰컨트롤러가 delegate 프로토콜을 강한 참조로 소유하고 있었고, 2. 뷰컨트롤러 내에서 사용된 일부 클로저에서 self를 강한참조해서 였습니다.
  1. 뷰컨트롤러 <-> delegate프로토콜
      - 뷰컨트롤러는 delegate를 강하게 참조하고, delegate는 옵셔널로 초기화되고 인스턴스를 뷰컨트롤러 외부에서 주입되는 상황
      - 뷰컨트롤러를 pop할 때, delegate에 할당된 프로토콜을 메모리에서 해제하지 않음
      - delegate는 강한 참조로 선언되어 뷰컨트롤러가 사라져도 프로토콜 인스턴스가 메모리에 남아 있으며, 이를 해제할 방법이 없어 누수 발생
      - (해결) delegate를 weak var로 선언하고, delegate 프로토콜을 클래스 전용으로 만들어 해결
  2. 뷰컨트롤러 내부 클로저에서 self 강한 참조
      - 뷰컨트롤러 <-> 클로저 내의 self
      - 뷰컨트롤러는 클로저를 소유
      - 클로저는 자신 외부의 self(뷰컨트롤러) 를 캡처해 소유함
      - 뷰컨트롤러가 pop되더라도 뷰컨트롤러 인스턴스에 대한 클로저의 참조 카운트가 남아 있기 때문에 메모리에서 해제되지 않아 누수 발생
      - (해결) 클로저 내에서 self를 사용 시 weak self로 선언해 해결

### 로그인화면 → 로그인 완료 → 앱으로 돌아왔을 때 로그인 화면에서 멈춰 있는 현상 디버깅
다른 디버깅 및 리팩토링 작업을 하다 발견한 오류입니다.
> 앱 설치 후 OAuth로그인을 하는 시점에 권한을 승인했는데도 앱으로 돌아와 다음으로 보여줘야 하는 저장소 목록 화면이 보여지지 않았고, 뜬금없이 `Expected to decode Array<Any> but found a dictionary instead.` 라는 Alamofire decoding Type Missmatch error가 발생하며 화면이 멈추는 현상.
- 저장소 목록을 받아오기 위해 `[Repository]`타입의 응답을 기대하는데 왜 딕셔너리로 응답이 왔는지? - 깃헙 API 문서를 찾아봐도 이에 대한 설명은 없었음
- 되는 부분부터 차례로 디버깅을 진행하니 IssueService를 초기화할 때 아주 옛날에 사용한 `IssueService(token: 토큰 ?? "")` 코드가 문제였음.
  - '앱을 처음 설치 후 실행'하는 시점에는 토큰이 없으며 이 때문에 IssueService의 토큰은 ""로 초기화되어버림
  - 이후 OAuth권한을 받고 액세스토큰을 가져왔으나, 가져온 토큰을 새롭게 설정하는 코드가 없어 계속 토큰값으로 "" 를 가지고 있었음
  - 저장소 목록을 요청할 때 액세스토큰값으로 ""를 가지고 요청함
  - 이 때 잘못된 토큰값으로 요청하면 `[Repostiory]`가 아닌 에러코드와 함께 JSON 딕셔너리 형식으로 응답이 오며, 이것을 받아 TypeMissmatch 에러가 발생했던 것..
- (해결) 액세스토큰을 얻은 이후 메서드를 통해 IssueService의 토큰값을 설정하고, IssueService의 토큰 변수를 옵셔널로 타입 변경
프로젝트가 매우 초반이었을 때 문제가 되는 코드를 작성했고, 이에 대해 한번 언급해주셨던 것 같은데 이렇게 큰 스노우볼이 되어 돌아올 줄은 몰랐습니다..😨 앞으로 옵셔널 기본값 지정은 더욱 주의해야겠다는 교훈을 얻었습니다.

---

저는 이번 PR을 마지막으로 IssueTracker 프로젝트는 마무리 작업을 하고, wiki에 배운 내용을 정리하려고 합니다!
오랜 시간 동안 함께 IssueTracker 프로젝트가 나아가게 도와 주시고, 세심하게 코드리뷰 해 주셔서 정말정말 감사드립니다.🙇‍♀️
리뷰어님 덕분에 많은 것을 배울 수 있었던 프로젝트였습니다!
여기에서 배웠던 지식과 교훈을 발판 삼아 다음에는 더 멋진 프로젝트 만들어 보겠습니다. 감사합니다🥹🙏👍

GangWoon commented:
우선 정말 수고많으셨습니다. 프로젝트를 유지보수하시면서 생각보다 많은 문제에 대해서 고민하셨던거 같습니다. 고민하셨던 점에대해서 리드미에 정리하시면 너무좋을거같습니다.


1. 모든 클로저가 약한 참조를 할 필요는 없습니다. 언제 약한참조가 필요한지 명확한기준을 공부하셨으면 좋겠습니다.
2. Enum을 사용해서 관리를 하고 있는 부분을 인상 깊게 봤습니다. 원하는 데이터를 로우 벨류로 가공하기 위해서 중간 레이어를 어떤 방식으로 설계할지 체험할 수 있는 좋은 경험 이었던거 같습니다.
3. 에러가 발생할 수 있는 상황에 대해서 민감하게 생각하셨으면 좋겠습니다. Throws라는 키워드를 사용해서 안전한 코드를 작성하셨으면 좋겠습니다.
4. Generic type에 대해서 살펴보셨으면좋겠습니다. 함수 외부에서 필요한 정보를 알려줄 수있다면 형변환을 매번할 필요가 사라질거 같습니다.

프로젝트를 꼼꼼하게 관리한다는 느낌을 받았습니다. 사용하지 않는 주석도 확인해가면서 지우시는 모습이 너무 좋았습니다. 접근자에 대해서도 객체입장을 한번더 생각하시는 습관은 추후 읽는 사람에게 많은 도움을 줄겁니다. 무엇보다도 기존 프로젝트를 유지보수하는 작업 이 쉽지 않으셨을텐데 정말 수고많셨습니다. 앞으로도 즐거운 마음으로 개발하실 수 있기를 응원합니다.
